### PR TITLE
scheduler module for replication

### PR DESCRIPTION
**Pull request by yuanshuhan**
submit the scheduler module for replication

### CODE REVIEW

File: src/replication/ng/scheduler/schedule.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		task.SrcResource = string(src)
+		dest, err := json.Marshal(destResource)
+		if err != nil {
+			log.Errorf("failed to marshal the destResource of %v.err:%s!", destResource, err.Error())
+			task.Status = "Error"
+			tasks = append(tasks, task)
+			continue
+		}
+		task.DstResource = string(dest)
+		task.Status = "Initial"
```
Comment (mmpei) [inline src/replication/ng/scheduler/schedule.go]: we don't design a Initial status for task, maybe should use Pending?
i feel a little confused about the translate of the status.
init(Pending)->js accept and return a jobID(running)->run success(Success)
                                                                               ->run failed(Failed)
                                                                               ->retrying(Retrying) this not support now
                                                                               ->stoped(Stopped)
@steven-zou @ywk253100

File: src/replication/ng/scheduler/schedule.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+		job.Name = common_job.ImageTransfer
+		job.Parameters = map[string]interface{}{
+			"src_resource": task.SrcResource,
+			"dst_resource": task.DstResource,
+		}
+		uuid, err := d.client.SubmitJob(job)
+		if err != nil {
+			log.Errorf("failed to submit the task:%v .err:%s!", task, err.Error())
+			task.Status = "Error"
```
Comment (mmpei) [inline src/replication/ng/scheduler/schedule.go]: this status should matched with the status defined in execution.go(wenkai created).
	TaskStatusFailed     string = "Failed"
	TaskStatusSucceed    string = "Succeed"
	TaskStatusStopped    string = "Stopped"
	TaskStatusInProgress string = "InProgress"
	TaskStatusPending    string = "Pending"

File: src/replication/ng/scheduler/schedule.go (hunk 1 above)
Comment (yuanshuhan) [reply src/replication/ng/scheduler/schedule.go]: remove the task status from scheduler, leave the status management to flow controller

File: src/replication/ng/scheduler/schedule.go (hunk 2 above)
Comment (yuanshuhan) [reply src/replication/ng/scheduler/schedule.go]: remove the task status from scheduler, leave the status management to flow controller

File: src/replication/ng/scheduler/scheduler.go [hunk 3]
```diff
@@ -42,3 +64,85 @@ type Scheduler interface {
 	// Stop the job specified by ID
 	Stop(id string) error
 }
+
+// Preprocess the resources and returns the item list that can be scheduled
+func (d *DefaultReplicator) Preprocess(srcResources []*model.Resource, destResources []*model.Resource) ([]*ScheduleItem, error) {
+	if len(srcResources) != len(destResources) {
+		err := errors.New("srcResources has different length with destResources")
```
Comment (steven-zou) [inline src/replication/ng/scheduler/scheduler.go]: if this error will be bubbled to the users finally, I suggest we should use some more meaningful statements to describe the problem.

File: src/replication/ng/scheduler/scheduler.go [hunk 4]
```diff
@@ -42,3 +64,85 @@ type Scheduler interface {
...
+}
+
+// Schedule transfer the tasks to jobs,and then submit these jobs to job service.
+func (d *DefaultReplicator) Schedule(items []*ScheduleItem) ([]*ScheduleResult, error) {
+	var results []*ScheduleResult
+	for _, item := range items {
+		if item.TaskID == 0 {
+			err := errors.New("some tasks do not have a ID")
+			log.Errorf(err.Error())
+			return nil, err
```
Comment (ywk253100) [inline src/replication/ng/scheduler/scheduler.go]: Put the error to its result, don't mark the whole function failed, or the flow controller will mark the whole execution failed.

File: src/replication/ng/scheduler/scheduler.go [hunk 5]
```diff
@@ -42,3 +64,85 @@ type Scheduler interface {
 	// Stop the job specified by ID
 	Stop(id string) error
 }
+
+// Preprocess the resources and returns the item list that can be scheduled
+func (d *DefaultReplicator) Preprocess(srcResources []*model.Resource, destResources []*model.Resource) ([]*ScheduleItem, error) {
+	if len(srcResources) != len(destResources) {
+		err := errors.New("srcResources has different length with destResources")
+		log.Errorf(err.Error())
```
Comment (ywk253100) [inline src/replication/ng/scheduler/scheduler.go]: Don't log error, just return it, the caller is responsible for that. The same to other error handle logic

**Comment by ywk253100** [comment]
@yuanshuhan Please squash your commits, thanks.

**Pull request merged**